File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/ItemSyncAdapter.java
package eu.pretix.libpretixsync.sync;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Iterator;
import java.util.List;

import eu.pretix.libpretixsync.api.ApiException;
import eu.pretix.libpretixsync.api.PretixApi;
import eu.pretix.libpretixsync.db.Item;
import eu.pretix.libpretixsync.db.Migrations;
import eu.pretix.libpretixsync.utils.HashUtils;
import eu.pretix.libpretixsync.utils.JSONUtils;
import io.requery.BlockingEntityStore;
import io.requery.Persistable;
import io.requery.query.Tuple;
import io.requery.util.CloseableIterator;

public class ItemSyncAdapter extends BaseConditionalSyncAdapter<Item, Long> {

    public ItemSyncAdapter(BlockingEntityStore<Persistable> store, FileStorage fileStorage, String eventSlug, PretixApi api, String syncCycleId, SyncManager.ProgressFeedback feedback) {
        super(store, fileStorage, eventSlug, api, syncCycleId, feedback);
    }

    @Override
    public void updateObject(Item obj, JSONObject jsonobj) throws JSONException {
        obj.setEvent_slug(eventSlug);
        obj.setServer_id(jsonobj.getLong("id"));
        obj.setPosition(jsonobj.getLong("position"));
        obj.setCategory_id(jsonobj.optLong("category"));
        obj.setAdmission(jsonobj.optBoolean("admission", false));
        obj.setActive(jsonobj.optBoolean("active", true));
        obj.setJson_data(jsonobj.toString());

        String remote_filename = jsonobj.optString("picture");
        if (remote_filename != null && remote_filename.startsWith("http")) {
            String hash = HashUtils.toSHA1(remote_filename.getBytes());
            String local_filename = "item_" + obj.getServer_id() + "_" + hash + remote_filename.substring(remote_filename.lastIndexOf("."));
            if (obj.getPicture_filename() != null && !obj.getPicture_filename().equals(local_filename)) {
                fileStorage.delete(obj.getPicture_filename());
                obj.setPicture_filename(null);
            }
            if (!fileStorage.contains(local_filename)) {
                try {
                    PretixApi.ApiResponse file = api.downloadFile(remote_filename);
                    OutputStream os = fileStorage.writeStream(local_filename);
                    InputStream is = file.getResponse().body().byteStream();
                    byte[] buffer = new byte[1444];
                    int byteread;
                    while ((byteread = is.read(buffer)) != -1) {
                        os.write(buffer, 0, byteread);
                    }
                    is.close();
                    os.close();
                    obj.setPicture_filename(local_filename);
                } catch (ApiException e) {
                    // TODO: What to do?
                    e.printStackTrace();
                } catch (IOException e) {
                    // TODO: What to do?
                    e.printStackTrace();
                    fileStorage.delete(local_filename);
                }
            } else {
                obj.setPicture_filename(local_filename);
            }
        } else {
            if (obj.getPicture_filename() != null) {
                fileStorage.delete(obj.getPicture_filename());
                obj.setPicture_filename(null);
            }
        }
    }

    @Override
    public CloseableIterator<Item> runBatch(List<Long> ids) {
        return store.select(Item.class)
                .where(Item.EVENT_SLUG.eq(eventSlug))
                .and(Item.SERVER_ID.in(ids))
                .get().iterator();
    }

    @Override
    CloseableIterator<Tuple> getKnownIDsIterator() {
        return store.select(Item.SERVER_ID)
                .where(Item.EVENT_SLUG.eq(eventSlug))
                .get().iterator();
    }

    @Override
    String getResourceName() {
        return "items";
    }

    @Override
    Long getId(JSONObject obj) throws JSONException {
        return obj.getLong("id");
    }

    @Override
    Long getId(Item obj) {
        return obj.getServer_id();
    }

    @Override
    Item newEmptyObject() {
        return new Item();
    }

    public void standaloneRefreshFromJSON(JSONObject data) throws JSONException {
        Item obj = store.select(Item.class)
                .where(Item.SERVER_ID.eq(data.getLong("id")))
                .get().firstOr(newEmptyObject());
        JSONObject old = null;
        if (obj.getId() != null) {
            old = obj.getJSON();
        }

        // Store object
        data.put("__libpretixsync_dbversion", Migrations.CURRENT_VERSION);
        data.put("__libpretixsync_syncCycleId", syncCycleId);
        if (old == null) {
            updateObject(obj, data);
            store.insert(obj);
        } else {
            if (!JSONUtils.similar(data, old)) {
                updateObject(obj, data);
                store.update(obj);
            }
        }
    }
}

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/ItemSyncAdapter.kt
package eu.pretix.libpretixsync.sync

import app.cash.sqldelight.TransactionWithoutReturn
import app.cash.sqldelight.db.QueryResult
import eu.pretix.libpretixsync.api.ApiException
import eu.pretix.libpretixsync.api.PretixApi
import eu.pretix.libpretixsync.db.Migrations
import eu.pretix.libpretixsync.sqldelight.Item
import eu.pretix.libpretixsync.sqldelight.SyncDatabase
import eu.pretix.libpretixsync.sync.SyncManager.ProgressFeedback
import eu.pretix.libpretixsync.utils.HashUtils
import eu.pretix.libpretixsync.utils.JSONUtils
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.io.OutputStream

class ItemSyncAdapter(
    db: SyncDatabase,
    fileStorage: FileStorage,
    eventSlug: String,
    api: PretixApi,
    syncCycleId: String,
    feedback: ProgressFeedback?,
) : BaseConditionalSyncAdapter<Item, Long>(
    db = db,
    fileStorage = fileStorage,
    eventSlug = eventSlug,
    api = api,
    syncCycleId = syncCycleId,
    feedback = feedback,
) {

    override fun getResourceName(): String = "items"

    override fun getId(obj: Item): Long = obj.server_id!!

    override fun getId(obj: JSONObject): Long = obj.getLong("id")

    override fun getJSON(obj: Item): JSONObject = JSONObject(obj.json_data)

    override fun queryKnownIDs(): MutableSet<Long> {
        val res = mutableSetOf<Long>()
        db.itemQueries.selectServerIdsByEventSlug(eventSlug).execute { cursor ->
            while (cursor.next().value) {
                val id = cursor.getLong(0)
                    ?: throw RuntimeException("server_id column not available")

                res.add(id)
            }
            QueryResult.Unit
        }

        return res
    }

    override fun insert(jsonobj: JSONObject) {
        val pictureFilename = processPicture(jsonobj, null)

        db.itemQueries.insert(
            active = jsonobj.optBoolean("active", true),
            admission = jsonobj.optBoolean("admission", false),
            category_id = jsonobj.optLong("category"),
            checkin_text = jsonobj.optString("checkin_text"),
            event_slug = eventSlug,
            json_data = jsonobj.toString(),
            picture_filename = pictureFilename,
            position = jsonobj.getLong("position"),
            server_id = jsonobj.getLong("id"),
            ticket_layout_id = null,
            ticket_layout_pretixpos_id = null,
        )
    }

    override fun update(obj: Item, jsonobj: JSONObject) {
        val pictureFilename = processPicture(jsonobj, obj.picture_filename)

        db.itemQueries.updateFromJson(
            active = jsonobj.optBoolean("active", true),
            admission = jsonobj.optBoolean("admission", false),
            category_id = jsonobj.optLong("category"),
            checkin_text = jsonobj.optString("checkin_text"),
            event_slug = eventSlug,
            json_data = jsonobj.toString(),
            picture_filename = pictureFilename,
            position = jsonobj.getLong("position"),
            id = obj.id,
        )
    }

    override fun delete(key: Long) {
        db.itemQueries.deleteByServerId(key)
    }

    private fun processPicture(jsonobj: JSONObject, oldFilename: String?): String? {
        val remote_filename: String = jsonobj.optString("picture")
        var result: String? = null;

        if (remote_filename.startsWith("http")) {
            val hash = HashUtils.toSHA1(remote_filename.toByteArray())
            val local_filename =
                "item_" + jsonobj.getLong("id") + "_" + hash + remote_filename.substring(
                    remote_filename.lastIndexOf(".")
                )
            if (oldFilename != null && oldFilename != local_filename) {
                fileStorage.delete(oldFilename)
                result = null
            }
            if (!fileStorage.contains(local_filename)) {
                try {
                    val file = api.downloadFile(remote_filename)

                    val outStream: OutputStream = fileStorage.writeStream(local_filename)
                    val inStream = file.response.body?.byteStream()

                    if (inStream == null) {
                        outStream.close()
                        throw IOException()
                    }

                    val buffer = ByteArray(1444)
                    var byteread: Int
                    while (inStream.read(buffer).also { byteread = it } != -1) {
                        outStream.write(buffer, 0, byteread)
                    }
                    inStream.close()
                    outStream.close()
                    result = local_filename
                } catch (e: ApiException) {
                    // TODO: What to do?
                    e.printStackTrace()
                } catch (e: IOException) {
                    // TODO: What to do?
                    e.printStackTrace()
                    fileStorage.delete(local_filename)
                }
            } else {
                result = local_filename
            }
        } else {
            if (oldFilename != null) {
                fileStorage.delete(oldFilename)
                result = null
            }
        }

        return result
    }

    override fun runInTransaction(body: TransactionWithoutReturn.() -> Unit) {
        db.itemQueries.transaction(false, body)
    }

    override fun runBatch(parameterBatch: List<Long>): List<Item> =
        db.itemQueries.selectByServerIdListAndEventSlug(
            server_id = parameterBatch,
            event_slug = eventSlug,
        ).executeAsList()

    @Throws(JSONException::class)
    fun standaloneRefreshFromJSON(data: JSONObject) {
        val obj = db.itemQueries.selectByServerId(data.getLong("id")).executeAsOneOrNull()
        val old: JSONObject? = obj?.json_data?.let { JSONObject(it) }

        // Store object
        data.put("__libpretixsync_dbversion", Migrations.CURRENT_VERSION)
        data.put("__libpretixsync_syncCycleId", syncCycleId)
        if (old == null) {
            insert(data)
        } else {
            if (!JSONUtils.similar(data, old)) {
                update(obj, data)
            }
        }
    }
}
